### PR TITLE
MRG: data loader functions

### PR DESCRIPTION
**Pull request by changsiyao**
#16 
Wrote load_interval and load_movement functions.
Please check.
Also, I'm not very sure how the `from mousestyle import data_dir` works. But I tried writing my path with `data_dir` joined in. If it's correct, I will just delete my original path without `data_dir` (they are commented out now).
@dylandaniels

### CODE REVIEW

File: mousestyles/data/__init__.py [hunk 1]
```diff
@@ -62,3 +62,98 @@ def load_all_features():
...
+    feature: str
+             where feature is one of ["AS", "F", "IS", "M_AS", "M_IS", "W"]
+    Returns
+    -------
+    output type: DataFrame
+    description: All the data from the feature directory into one data frame
+                 columns are strain, mouse, day, start, stop
+    """
+    # get all file names
+    # file_names = _os.listdir("intervals/" + feature)
```
Comment (peterasujan) [inline mousestyles/data/__init__.py]: Should probably clean up commented out lines of code before merging.

File: mousestyles/data/__init__.py [hunk 2]
```diff
@@ -62,3 +62,98 @@ def load_all_features():
...
+    # CT = np.load("txy_coords/CT/CT_strain" + str(strain) +
+    #     "_mouse" + str(mouse) + "_day" + str(day) + ".npy")
+    # CX = np.load("txy_coords/CX/CX_strain" + str(strain) +
+    #     "_mouse" + str(mouse) + "_day" + str(day) + ".npy")
+    # CY = np.load("txy_coords/CY/CY_strain" + str(strain) +
+    #     "_mouse" + str(mouse) + "_day" + str(day) + ".npy")
+    HB = np.load(
+        _os.path.join(
+            data_dir,
+            "txy_coords/C_idx_HB/C_idx_HB_strain" + str(strain) +
```
Comment (peterasujan) [inline mousestyles/data/__init__.py]: You might consider changing these to use [format strings](https://docs.python.org/3.5/library/string.html#formatstrings), which are in my opinion more readable.

**Comment by peterasujan** [comment]
@changsiyao this looks great! I just left a couple minor line comments.

**Comment by changsiyao** [comment]
@peterasujan by using format string, do you mean something like this?
"txy_coords/CX/CX_strain%s_mouse%s_day%s.npy" % (strain, mouse, day)

**Comment by changsiyao** [comment]
@peterasujan I modified the code as per your suggestions. Could you please check again? Thanks!

**Comment by peterasujan** [comment]
Exactly, looks sweet! I think this is ready to be changed to MRG, but perhaps @dylandaniels should take a look as well.

File: mousestyles/data/__init__.py [hunk 3]
```diff
@@ -62,3 +62,80 @@ def load_all_features():
...
+
+
+def load_intervals(feature):
+    """
+    Parameters
+    ----------
+    feature: str
+             where feature is one of ["AS", "F", "IS", "M_AS", "M_IS", "W"]
+    Returns
+    -------
```
Comment (dylandaniels) [inline mousestyles/data/__init__.py]: This isn't quite the docstring format. See: https://github.com/berkeley-stat222/mousestyles/blob/master/DOCFLOW.md

I think it should look like

intervals: pandas.DataFrame
     <blah blah description here>

The same goes for the other function.

File: mousestyles/data/__init__.py [hunk 4]
```diff
@@ -62,3 +62,80 @@ def load_all_features():
     other_features = [x.iloc[:, -1] for x in data_frames]
     other_features.insert(0, all_features_df)
     return pd.concat(other_features, axis=1)
+
+
+def load_intervals(feature):
+    """
+    Parameters
+    ----------
+    feature: str
```
Comment (dylandaniels) [inline mousestyles/data/__init__.py]: Choices should be put in brackets... see 
https://github.com/berkeley-stat222/mousestyles/blob/master/DOCFLOW.md

File: mousestyles/data/__init__.py [hunk 5]
```diff
@@ -62,3 +62,80 @@ def load_all_features():
...
+             where feature is one of ["AS", "F", "IS", "M_AS", "M_IS", "W"]
+    Returns
+    -------
+    output type: DataFrame
+    description: All the data from the feature directory into one data frame
+                 columns are strain, mouse, day, start, stop
+    """
+    # get all file names
+    file_names = _os.listdir(_os.path.join(data_dir, "intervals", feature))
+    file_names = [name for name in file_names if len(name) > 10]
```
Comment (dylandaniels) [inline mousestyles/data/__init__.py]: why len(name) > 10?

Comment (changsiyao) [inline mousestyles/data/__init__.py]: Oh, maybe because I downloaded the zip data, so when I unzip it, there were some hidden file thats not .npy data. So I add it that to ensure I'm reading all the files I want. But I can delete that part if its not necessary.
and what do you mean by choices should be in brackets?
@dylandaniels

**Comment by dylandaniels** [comment]
Awesome, @changsiyao! This looks great--thanks so much for going above and beyond and taking on both of these tasks. I have a couple comments I put inline above--mostly style things.

You may want to do an interactive rebase to squash all your commits into one commit before merging.

**Comment by changsiyao** [comment]
@dylandaniels I think I did all the corrections. Could you check again please? Thanks!

**Comment by jarrodmillman** [comment]
This is a nice example of the review process.  Feel free to merge; however, this might be a good example to show to the class.  If you are OK with using this pull request as an example, you could first walk the class through what happened and how things improved.  Then we could ask everyone to look at the current code for 5-10 minutes to see if people have any other suggestions and/or improvements

I have a couple of ideas for the code, but am not sure they should be pursued.  If we discuss this in class, I could suggest my ideas and we could discuss them in class.

If this sounds like a good idea, we could also show everyone an example of interactive rebasing and the merge process.

This could happen during the second half of class on Monday.  If you need to merge this sooner, go ahead.  Otherwise, I am happy to help present the demo / example of the review process on Monday with you.

**Comment by jarrodmillman** [comment]
I assigned this to @shamindras, since both @peterasujan and @dylandaniels have already reviewed it.  Just in case you all agree that this could be a good example for the class.

**Comment by shamindras** [comment]
@jarrodmillman - sure, sounds good! I can just prepare some review comments by Monday. And we can then add them in live from your computer and we can resubmit it for alterations.

If they are quick, then they could be done live as well so that everyone can see the end-to-end cycle of review-alteration-MRG

File: mousestyles/data/__init__.py [hunk 6]
```diff
@@ -62,3 +62,80 @@ def load_all_features():
     other_features = [x.iloc[:, -1] for x in data_frames]
     other_features.insert(0, all_features_df)
     return pd.concat(other_features, axis=1)
+
+
+def load_intervals(feature):
+    """
+    Parameters
+    ----------
+    feature: {"AS", "F", "IS", "M_AS", "M_IS", "W"}
```
Comment (shamindras) [inline mousestyles/data/__init__.py]: This should just be a **single** feature right i.e. **one** of the `{"AS", "F", "IS", "M_AS", "M_IS", "W"}` possible values. Also is this a _string input_ i.e. `"AS"` or a _set input_ `{"AS"}`

File: mousestyles/data/__init__.py [hunk 7]
```diff
@@ -62,3 +62,80 @@ def load_all_features():
...
+
+    Returns
+    -------
+    intervals : pandas.DataFrame
+        All the data from the feature directory into one data frame
+        columns are strain, mouse, day, start, stop
+    """
+    # get all file names
```
Comment (shamindras) [inline mousestyles/data/__init__.py]: We should add a line of exception error handling to ensure that only **one** of the `{"AS", "F", "IS", "M_AS", "M_IS", "W"}` is entered in the input and exit gracefully with a message like _ "input value must be one of {"AS", "F", "IS", "M_AS", "M_IS", "W"}"_.

Here is [a guide](http://hplgit.github.io/primer.html/doc/pub/input/._input-bootstrap007.html) to doing this.

File: mousestyles/data/__init__.py [hunk 8]
```diff
@@ -62,3 +62,80 @@ def load_all_features():
...
+    CT = np.load(_os.path.join(data_dir, CT_path))
+    CX = np.load(_os.path.join(data_dir, CX_path))
+    CY = np.load(_os.path.join(data_dir, CY_path))
+    # make data frame
+    dt = pd.DataFrame()
+    dt["t"] = CT
+    dt["x"] = CX
+    dt["y"] = CY
+    dt["isHB"] = HB
+    dt = dt.sort(["t"])
```
Comment (shamindras) [inline mousestyles/data/__init__.py]: Is the having the output Dataframe _sorted_ essential here? Can we remove this if unecesary. If sorting adds a trivial amount to the runtime (<1s) then you can probably leave it in.

File: mousestyles/data/__init__.py (hunk 6 above)
Comment (dylandaniels) [inline mousestyles/data/__init__.py]: This is what the documentation means, @shamindras. The convention for a choice is to put it in set brackets. I recommended this to @changsiyao above.

File: mousestyles/data/tests/test_data.py [hunk 9]
```diff
@@ -12,3 +12,13 @@
 def test_all_features_loader():
     all_features = data.load_all_features()
     assert_equal(all_features.shape, (21131, 13))
+
+
```
Comment (shamindras) [inline mousestyles/data/tests/test_data.py]: Could we add a one line comment to unit test i.e. _#Checking dimensions of test all features load dataset are correct after importing"_

File: mousestyles/data/tests/test_data.py [hunk 10]
```diff
@@ -12,3 +12,13 @@
...
+def test_intervals_loader():
```
Comment (shamindras) [inline mousestyles/data/tests/test_data.py]: Could we add a one line comment to unit test explaining what it does

File: mousestyles/data/tests/test_data.py [hunk 11]
```diff
@@ -12,3 +12,13 @@
...
+    AS = data.load_intervals('AS')
+    assert_equal(AS.shape, (1343, 5))
+
+
```
Comment (shamindras) [inline mousestyles/data/tests/test_data.py]: One line comment explaining what unit test does

File: mousestyles/data/__init__.py (hunk 7 above)
Comment (changsiyao) [inline mousestyles/data/__init__.py]: Then, should I also check if the strain, mouse, day values are nonnegative integers in the load_movement function?

File: mousestyles/data/__init__.py (hunk 6 above)
Comment (shamindras) [inline mousestyles/data/__init__.py]: I see @dylandaniels - so this means I run this function as `load_intervals("AS")`

File: mousestyles/data/__init__.py (hunk 7 above)
Comment (shamindras) [inline mousestyles/data/__init__.py]: @changsiyao - I believe we should have input validation and handling for all of our functions and a unit test as well.

@dylandaniels - In this case has the validation been taken care of at some higher level before calling this function?

Comment (changsiyao) [inline mousestyles/data/__init__.py]: @shamindras Do you want me to write separate unit tests for checking inputs or just add an assert in the current functions?

Comment (shamindras) [inline mousestyles/data/__init__.py]: @changsiyao - perhaps a new separate function just for input validation e.g. `test_feature_load_input` and add multiple asserts there. This will modularise the type of tests out more

Comment (changsiyao) [inline mousestyles/data/__init__.py]: Also, I'm not sure how to test whether file_names is not an empty list. Any ideas? @shamindras

Comment (shamindras) [inline mousestyles/data/__init__.py]: Sure [this should help](http://stackoverflow.com/questions/16021096/python-best-way-to-check-if-list-is-empty-without-not)

The simple way is to use the `if not file_names:` syntax to check for emptiness. There is a subtle issue that this also picks up `None` objects. However we can check with Jarrod if and how to  best capture these.

Comment (changsiyao) [inline mousestyles/data/__init__.py]: @shamindras I actually meant how to write unit test for that, cuz currently no folder is empty, so I can't test whether I can catch the error of some directory being empty.

**Comment by changsiyao** [comment]
So, I was trying to see if the new checking input test would pass, but it failed because the `assert_raises` function needs to `import nose` and here is the error message `ImportError: Need nose >= 0.10.0 for tests - see http://somethingaboutorange.com/mrl/projects/nose`
Any ideas? @shamindras

File: mousestyles/data/__init__.py (hunk 6 above)
Comment (peterasujan) [inline mousestyles/data/__init__.py]: > so this means you should run this function as `load_intervals("AS")`?
@shamindras That's my understanding.
@changsiyao I think a couple short examples in the docstring would make things really clear, when you get a chance. You can just make another section `Examples` and then add a couple [doctests](https://docs.python.org/3.5/library/doctest.html) as examples.

**Comment by shamindras** [comment]
@jarrodmillman - I did a code review for this today. @changsiyao implemented all of the suggestions. Have left this as WIP so that you can perform your review tomorrow with valued suggestions.

This can be the final code review tutorial for this `PR` you wanted to demonstrate _in class_.

**Comment by jarrodmillman** [comment]
I will take responsibility for merging this pull request.  I would like to wait until class on Wednesday.  Please let me know if that is a problem.

I am going to ask everyone in the class to take a look at this.  I will also ask them to suggest improvements (including code snippets).  I have a coupe things I would like you to address as well.

File: mousestyles/data/__init__.py [hunk 12]
```diff
@@ -62,3 +62,105 @@ def load_all_features():
     other_features = [x.iloc[:, -1] for x in data_frames]
     other_features.insert(0, all_features_df)
     return pd.concat(other_features, axis=1)
+
+
+def load_intervals(feature):
+    """
```
Comment (jarrodmillman) [inline mousestyles/data/__init__.py]: Please add a short 1 sentence summary of each function at the top of the docstring.  While the function name should be as self-explanatory as possible, it is often a useful exercise to expand the encoded function name into a short sentence or phrase.  This can help you think about whether the name of the function is appropriate.  It also provides a bit more clarity to the authors intention, which is helpful for users and collaborators.

For example, 
"Return the project interval data for the specified feature."

File: mousestyles/data/__init__.py [hunk 13]
```diff
@@ -62,3 +62,105 @@ def load_all_features():
...
+        sub = np.load(path)
+        dt_sub = pd.DataFrame()
+        dt_sub["strain"] = [strain] * sub.shape[0]
+        dt_sub["mouse"] = [mouse] * sub.shape[0]
+        dt_sub["day"] = [day] * sub.shape[0]
+        dt_sub["start"] = sub[:, 0]
+        dt_sub["stop"] = sub[:, 1]
+        dt = pd.concat([dt, dt_sub])
+    # sort based on strain, mouse and day
+    dt = dt.sort(["strain", "mouse", "day"])
```
Comment (shamindras) [inline mousestyles/data/__init__.py]: Do we need to explicitly sort the dataframe in the output? If not please remove this line

Comment (changsiyao) [inline mousestyles/data/__init__.py]: I think it looks better when the return data frame is sorted already, since some of the analysis is base on mouse-day. I timed it and it takes less than 1 second to sort a 4260849 row data frame.

**Comment by changsiyao** [comment]
@jarrodmillman Can you check if we can merge this? The fails are due to the other code scripts from PR #82 running in the python 3 environment. Thanks!

**Comment by jarrodmillman** [comment]
This was rebased and failures were pre-existing (and caused by my pull request).  I will merge now.
